### ISSUE TITLE
Gairech void bug

### ISSUE THREAD

**Issue by Pegwald** [aura-project/aura#493]
This has happened twice now on my server that a player "fell" out of the world when they were around in Gairech.
One was caused when trying to enter Dunbarton from Gairech. The player clicked the "portal" to the Dunbarton area and fell out of the world.

The other one was fighting a wolf (I believe) around the dragon ruins and fell out to the same black void like the previous player also did.

**Comment by exectails** [comment]
As per usual, any warnings, errors, etc?

**Comment by exectails** [comment]
Could you ask them to use the `>where` command if it happens again, to see where they exactly they went to?

**Comment by Pegwald** [comment]
2016-11-29 10:29 [Warning] - CancelActiveSkill: Player 'Gray Wolf' tried to cancel a skill, without one being active.
2016-11-29 10:31 [Info] - Saved global script variables.
2016-11-29 10:32 [Debug] - Run locked for 'darkkir3'.
2016-11-29 10:36 [Info] - Saved global script variables.

2016-11-29 10:59 [Unimplemented] - PacketHandlerManager: Handler for '741B', '?'.
2016-11-29 10:59 [Debug] - Op: 0000741B ?, Id: 001000000000004A
001 [00A0000E000A0396] Long   : 45036056403903382

2016-11-29 10:59 [Unimplemented] - TouchProp: No prop behavior for '00A0000E000A0396'.
2016-11-29 10:59 [Unimplemented] - PacketHandlerManager: Handler for '741B', '?'.
2016-11-29 10:59 [Debug] - Op: 0000741B ?, Id: 001000000000004A
001 [00A0000E000A0396] Long   : 45036056403903382

2016-11-29 10:59 [Unimplemented] - TouchProp: No prop behavior for '00A0000E000A0396'.
2016-11-29 10:59 [Unimplemented] - PacketHandlerManager: Handler for '741B', '?'.
2016-11-29 10:59 [Debug] - Op: 0000741B ?, Id: 001000000000004A
001 [00A0000E000A0396] Long   : 45036056403903382

2016-11-29 10:59 [Unimplemented] - TouchProp: No prop behavior for '00A0000E000A0396'.
2016-11-29 10:59 [Unimplemented] - PacketHandlerManager: Handler for '741B', '?'.
2016-11-29 10:59 [Debug] - Op: 0000741B ?, Id: 001000000000004A
001 [00A0000E000A0396] Long   : 45036056403903382

2016-11-29 10:59 [Unimplemented] - TouchProp: No prop behavior for '00A0000E000A0396'.
2016-11-29 10:59 [Unimplemented] - PacketHandlerManager: Handler for '741B', '?'.
2016-11-29 10:59 [Debug] - Op: 0000741B ?, Id: 001000000000004A
001 [00A0000E000A0396] Long   : 45036056403903382

2016-11-29 10:59 [Unimplemented] - TouchProp: No prop behavior for '00A0000E000A0396'.
2016-11-29 10:59 [Unimplemented] - TouchProp: No prop behavior for '00A0000E000A02C5'.
2016-11-29 11:01 [Info] - Saved global script variables.
2016-11-29 11:03 [Warning] - StatMods.Add: Double stat mod for 'Equipment:22517998136860308'.
2016-11-29 11:04 [Unimplemented] - TouchProp: No prop behavior for '00A0000E00010014'.
2016-11-29 11:04 [Unimplemented] - TouchProp: No prop behavior for '00A0000E00010014'.
2016-11-29 11:04 [Unimplemented] - TouchProp: No prop behavior for '00A0000E00010014'.
2016-11-29 11:06 [Info] - Saved global script variables.

2016-11-29 11:16 [Info] - Saved global script variables.
2016-11-29 11:16 [Info] - 'darkkir3' is closing the connection. Saving...

That's all I managed to dig up :(

**Comment by Raraldor** [comment]
Oh, for the record I was fighting black wolves far east of Ciar dungeon, I killed on, clicked to switch targets to another wolf, and once I took maybe two steps I was teleported into the void.

**Comment by exectails** [comment]
Honestly, I'm starting to wonder if maybe there's a security hole somewhere, and someone is doing this *to* you, because I really don't understand how people are warped to coordinates like that all of a sudden, in completely different situations. That has never happened before, and nothing warp related changed recently as far as I know.

I could use a packet log of this happening, to see what the client and the server say to each other. If someone feels up to that task, checkout our logging tools:
http://aura-project.org/forum/topic/220-mabipale-packet-logger-updated-2015-12-20-v290/
http://aura-project.org/forum/topic/40-morrighan-client-proxy-updated-2015-12-13-v156/

Just log yourself playing, and when it happens, you stop the logger and send me the saved log file.

**Comment by exectails** [comment]
Also please keep the results of `>where` coming.

I'm kind of reluctant to make a hot-fix, so people don't just relog whenever it happens, and don't consider helping figure it out anymore^^"

**Comment by darkkir3** [comment]
We did a screen of the location: https://i.gyazo.com/97e70a0b11b3262398617da19e34abe9.png

**Comment by Raraldor** [comment]
On the topic of someone doing this to us, the server we are all playing on has an online list on it's website. There's a guy named Cyjuga who has been on during both the occasions  this happened, but is still level 1. When I try to add him as a friend it says unable to locate user. Is that just a gm Pegwald?

**Comment by Sairii** [comment]
that unfortunaly does not really help.
we need a log of that. teach pegwald how to do this and then let him packet log that or packet log it yourself.

**Comment by exectails** [comment]
>does not really help

It doesn't help a lot, but I did say keep the coordinates coming^^ And that on had a positive X coordinate, which makes it different from the other one o~o

**Comment by darkkir3** [comment]
Alright I managed to get stuck again (with intention of getting stuck of course) while logging my packets on the client side. Here is the packet log of the whole session:
http://filebin.ca/33nmvJ3cDld9/20161130_220036.txt

**Comment by Sairii** [comment]
Op: 0000659A, Id: 001000000000004A force run to???
exec wahatafa is that

**Comment by exectails** [comment]
Thank you, @darkkir3.

Alright, so we use ForceRunTo for exactly one thing: StopMove, basically forcing the character to go to its current position and stop there. (Officials do the same.) The question is how the server decides to make the character run to `int.MinValue`. I'm afraid I don't have a theory yet.

However, StopMove is used in various places, so it would at least explain the randomness.

**Comment by exectails** [comment]
One thing I can say so far is that you should be safe if you don't switch your weapons while you're moving for example. StopMove is only executed if you're actually moving during an action. Not that that's a solution, but maybe it can be reduced a little.

**Comment by exectails** [comment]
Can you tell me something about your latency? Do you have a good connection to the server? Any lag?

**Comment by Raraldor** [comment]
So loading magnum shot while running backward would cause it as well then, right?

**Comment by darkkir3** [comment]
> Can you tell me something about your latency? Do you have a good connection to the server? Any lag?

I am pretty sure that I am not lagging since im able to switch weapons in like half a sec

**Comment by exectails** [comment]
@Raraldor Yes, moving while loading skills where you come to a still stand could cause this I think. At least that's the theory =/

**Comment by Raraldor** [comment]
@exectails alright, thanks, I'll be sure to not do that lol

**Comment by exectails** [comment]
What confuses me is how this has never happened before, and suddenly it seems to be common among a few players. The position -1/-1 is also a curious one, it doesn't seem like a random glitch, it's a very specific position you apparently keep getting warped to.

**Comment by darkkir3** [comment]
> What confuses me is how this has never happened before, and suddenly it seems to be common among a few players. The position -1/-1 is also a curious one, it doesn't seem like a random glitch, it's a very specific position you apparently keep getting warped to.

Though I dont know much about the content of the packets used by mabi im just browsing through the latest build and looking for some obvious mistakes. Should be a simple error somewhere since it seems to be a static position.

**Comment by Sairii** [comment]
i cannot reproduce this :<
my fingers hurt from weapon switching

**Comment by exectails** [comment]
If it is a simple error, I haven't found it yet. Right now I'm assuming it's some obscure glitch in the movement calculations =/

**Comment by Raraldor** [comment]
It took dark about 15 minutes to reproduce it, and it doesn't seem to be something that happens all the time, or else we'd have everyone on the server trapped I'd imagine.

**Comment by darkkir3** [comment]
> i cannot reproduce this :<
my fingers hurt from weapon switching

Im playing osu since few years and when trying to invoke this bug i basically spammed that button like there was no tomorrow. All that finger practice finally got to use kappa

**Comment by Sairii** [comment]
my macro is doing its best.
did you do anything else? or just spam that key near dragon moon gate?

**Comment by darkkir3** [comment]
> my macro is doing its best.
did you do anything else? or just spam that key near dragon moon gate?

Well, I basically tried to do all the random things we previously got stuck for: Spamming the map change between dunbi and gairech hills, attacking random mobs, hitting props, interacting with empty props and other weird stuff. I was surprised that this happened while I was actually only walking on the field not doing anything weird besides switching the equipped weapons.

> If it is a simple error, I haven't found it yet. Right now I'm assuming it's some obscure glitch in the movement calculations =/

Code seems clean for me. It has to be something about this part:

Creature.cs:
`public override Position GetPosition()
		{
			if (!this.IsMoving)
				return _position;

			var passed = (DateTime.Now - _moveStartTime).TotalSeconds;
			if (passed >= _moveDuration)
				return this.SetPosition(_destination.X, _destination.Y);

			var xt = _position.X + (_movementX * passed);
			var yt = _position.Y + (_movementY * passed);

			return new Position((int)xt, (int)yt);
		}

`
The calculation itself seems to be fine. However this seems to be the only part where an issue could occur, since everything else seems alright (I repeat: I dont know the content of mabi packets, so i didnt check all the network classes). Now if i could only set up a test server on localhost to debug this :7

**Comment by exectails** [comment]
>Now if i could only set up a test server on localhost to debug this :7

What's stopping you?^^

Someone who's able to reproduce this needs to go into the code, figure out where exactly these values are coming from =/

**Comment by darkkir3** [comment]
> Someone who's able to reproduce this needs to go into the code, figure out where exactly these values are coming from =/

Yeah thats what I meant when I said to setup a debug server. I will setup a server today (evening) and investigate further.

**Comment by exectails** [comment]
Yea, I know that's what you meant^^ Thank you. I've tried reproducing it myself, but all I got was tiny jumps, nothing like a jump to -1/-1.

https://www.youtube.com/watch?v=Tgvz8h-LSeA

**Comment by darkkir3** [comment]
xt: NaN | yt: NaN
deltaTime: 0.9956776
MoveDuration: NaN
MoveStartTime: 12/1/2016 8:03:17 PM
posX: -2147483648 | posY: -2147483648
movementX: NaN  | movementY: NaN
destionationX: 31516 | destinationY: 99705
[Info] - [End]passed: 1.0001805
xt: NaN | yt: NaN
deltaTime: 1.0001805
posX: -2147483648 | posY: -2147483648
MoveDuration: NaN
MoveStartTime: 12/1/2016 8:03:17 PM
movementX: NaN  | movementY: NaN
destionationX: -2147483648 | destinationY: -2147483648

Kappa. Knew it had something to do with movement calculations owo Trying to figure out the issue in the code now. For some reason, MoveDuration is NaN.

**Comment by exectails** [comment]
>It will teleport you without any reason randomly.

What do you mean, just while walking around? These functions are several years old, how is it possible that they suddenly, "randomly" warp people around? =|

We can see that the result from GetPosition seems to be wrong. The following Move packet might then teleport the player to that position and continue to the destination from there. Because of that it would make sense for the problem to be in GetPosition, but I don't see what it could be... 

I guess we need logs from GetPosition as well, to see how exactly the incorrect position comes to be. Is it the SetPosition call, based on _destination? Or is the result of the calculation incorrect somehow? And if it destination, how is it set to some negative number?

**Comment by darkkir3** [comment]
> What do you mean, just while walking around? These functions are several years old, how is it possible that they suddenly, "randomly" warp people around? =|

I did a basic check against negative -100k or below on each method that sets the position but that didnt get triggered too. Honestly I am very confused myself rn :7

**Comment by exectails** [comment]
It seems most likely that something is set during the calculations, which messes them up somehow. I don't know what, or how, but if it happened just out of nowhere, randomly, I think we would've encountered this much earlier.

**Comment by exectails** [comment]
Hm... I wonder if I can reproduce it if I replay the packet log.

**Comment by darkkir3** [comment]
> Hm... I wonder if I can reproduce it if I replay the packet log.

It is pretty easy to reproduce by just walking around, attacking mobs and getting attacked. Takes around 15 mintues to 2 hours to trigger once though.

**Comment by darkkir3** [comment]
> I think we would've encountered this much earlier.

It can not be a coincidence that 3 different people get stuck around the same day. All those people didnt use any packet-modifying mods, so it has to be a server sided bug.

**Comment by darkkir3** [comment]
Found another bug (most likely not related to this one): Using custom speed values while shift walking will trick the server into thinking you are faster than you actually are (Since >speed only sets the running speed).

**Comment by QewQew** [comment]
If it's unrelated, then make a new issue.

**Comment by darkkir3** [comment]
The only way for moveDuration to be zero is by Creature.GetSpeed() returning zero. However this makes no sense whatsoever since those are constant values. Im done :7

**Comment by exectails** [comment]
>It is pretty easy to reproduce by just walking around

That's the thing though, people have played around with Aura for *years* now, and the movement code basically hasn't changed at all since the beginning, yet we've never had this issue. I'm not saying that you're modifying packets or anything, but this problem is mysterious, and apparently harder to reproduce than you make it look, seeing how seemingly nobody else was able to reproduce it yet =/

Did you stick any more debug logs into the code, to see what GetPosition looks like for example?

**Comment by exectails** [comment]
Just realized playing back the log won't work because the walk packets were filtered =/

**Comment by YaibaToKen** [comment]
I got sent to the void as well today xD twice, in sen mag plains, near the gairech hills warp...
>where:
Region 53 @ -2147483648 / -2147483648, Area: 0, Dir: 160 (Radians: 3,942)
Hope this helps

**Comment by exectails** [comment]
Is anyone who can *reproduce* this at will still here? =/

**Comment by YaibaToKen** [comment]
I honestly don't know but seeing as I'm grinding in the area I'll keep posting any info I find

**Comment by exectails** [comment]
I appreciate it, but the problem is that the results of `>where` don't get us anywhere anymore at this point^^ I'll have to add debug logs to relevant portions of the position calculation, so that when someone stumbles upon this bug we might get new information. We have to figure out where it comes from.

**Comment by YaibaToKen** [comment]
just wondering but, are maps even supposed to have minus coordinates?

**Comment by exectails** [comment]
I don't know of any maps that have accessible areas below 0,0, no.

**Comment by darkkir3** [comment]
@exectails: Create a debug build where pretty much everything gets logged. (And not only movement values, since that didnt help last time). You should know better where there is a possibilty for the char pos to change at all. Post that on here and i will be gladly helping by trying to trigger that crap. I wont search through the whole project to add debug logs myself, not motivated enough for that.

**Comment by exectails** [comment]
@darkkir3 I didn't expect you to, I was hoping for you to come back so we can discuss where to put the logs, because you're the only one so far to be able to reproduce it xD

You put logs into Move, but as we can see in that output you posted, the problem is not in Move, the position you get from GetPosition is already incorrect. This leads me to believe that the problem is in there. I'd like to ask you to do the same thing you did before, but in GetPosition.

```csharp
public override Position GetPosition()
{
	if (!this.IsMoving)
		return _position;

	var passed = (DateTime.Now - _moveStartTime).TotalSeconds;
	if (passed >= _moveDuration)
		return this.SetPosition(_destination.X, _destination.Y);

	var xt = _position.X + (_movementX * passed);
	var yt = _position.Y + (_movementY * passed);

	Log.Debug("passed: " + passed);
	Log.Debug("_moveDuration: " + _moveDuration);
	Log.Debug("_position: " + _position);
	Log.Debug("_movementX: " + _movementX);
	Log.Debug("_movementY: " + _movementY);
	Log.Debug("xt: " + xt);
	Log.Debug("yt: " + yt);

	return new Position((int)xt, (int)yt);
}
```

**Comment by darkkir3** [comment]
K seems like i'm to dumb to read. Was already a bit salty that day since I was literally trying to invoke this once in 2 hours. And when it actually triggered, i logged the wrong values. So i was like “meh, what a waste of time“.  Gonna debug this today after getting home from work.

**Comment by exectails** [comment]
Thank you =)

**Comment by xeroplz** [comment]
By the way, if you want to reproduce this at will, just run "out of bounds" in a map and switch weapons. You'll probably end up at 0,0.

**Comment by darkkir3** [comment]
> By the way, if you want to reproduce this at wil
Had problems using my dev server as database(debian). Have to install a local db for this. Triggering this isnt the hard part.

**Comment by darkkir3** [comment]
Just saying... Copied a brand new build. Dropped the whole database. Everything went fine. Channelserver took a bit to get up. After encountering any instance of EntityCreature - this happened:

`[Info] - Loading localization (en-US)...
[Info] - Initializing world...
[Info] -   done loading 407 regions.
[Info] -   done loading 0 guilds.
[Info] - Loading skill handlers...
[Info] -   done loading 63 skill handlers.
[Info] - Loading scripts, this might take a moment...
[Info] - Saved global script variables.
[Info] -   loaded 526 scripts (0 failed).
[Info] - Initializing scripts...
[Info] -   script initialization complete
[Info] - Spawning creatures...
[Info] -   done spawning creatures.
[Status] - Server ready, listening on 0.0.0.0:11020.

[Info] - Trying to connect to login server at 127.0.0.1:11000...
[Info] - Connection to login server at '127.0.0.1:11000' established.

[Info] - Type 'help' for a list of console commands.
[Info] - Connection established from '127.0.0.1:53842.
[Info] - Saved global script variables.
[Warning] - AiScript.Heartbeat: ChickenAi stuck?
[Warning] - AiScript.Heartbeat: SheepAi stuck?
[Warning] - AiScript.Heartbeat: SheepAi stuck?
d[Warning] - AiScript.Heartbeat: RaccoonAi stuck?
[Warning] - AiScript.Heartbeat: ChickenAi stuck?
[Warning] - AiScript.Heartbeat: ChickenAi stuck?
[Warning] - AiScript.Heartbeat: HenAi stuck?
[Warning] - AiScript.Heartbeat: RaccoonAi stuck?
[Warning] - AiScript.Heartbeat: HenAi stuck?
[Warning] - AiScript.Heartbeat: SheepAi `stuck?``

Seems like posting the position of every creature in the logs isnt a good idea. Gonna override this for player instances only.

**Comment by exectails** [comment]
Right, that's a good idea. Didn't think of that. You could also make it conditional, just for you.
```csharp
if (this.Name == "yourname")
```

**Comment by exectails** [comment]
@darkkir3 Any luck?
